### ISSUE TITLE
MutableObservableArray causes side effect reloadData

### ISSUE THREAD

**Issue by killev** [DeclarativeHub/Bond#636]
Hi guys,

I use **MutableObservableArray / .diff()** to simplify the logic and reduce number of operation on collectionView/tableView. But latest version brought new unexpected issue: it calls reloadData when there're no changes (even though this is even very expensive operation as it reloads all visible cells).

Take a look at the test which explains the issue. I expect not to get any event in this case except the initial reloadData. 

```swift 
    func testBatchUpdatesEmptyBatch() {
        array.batchUpdate { (array) in }

         let possibleResultA = [OrderedCollectionDiff<IndexPath>()] // It should call reloadData  once only

        XCTAssert(collectionView.observedEvents == possibleResultA)
    }
```

the cause of the issue is here:

```swift
        if diff.isEmpty {
            collectionView.reloadData()
        } else if diff.count == 1 {
...
```

There's no difference between "Empty Diff" or diff which should reload all data. But it should be.

**Comment by tonyarnold** [comment]
@killev is `batchUpdate(…)` the only place where this issue is reproducible?

It might be appropriate to add logic to that method to determine if any change occurred.

**Comment by killev** [comment]
Hi guys,

In general I disagree that empty diff should reload all tree.
Ideologically, it’s a bad practice when zero = all. Providing isreload
doesn’t introduce new state. This state already exists, but hidden.

The only thing is diff should be enum, not struct as isempty, isreload,
isperformupdates non-intersecting.

Sorry, I’m writing from iPhone and can be at laptop in couple of days to
provide more arguments why introducing this state is reasonable.

Peter

**Comment by tonyarnold** [comment]
Absolutely, take your time - I'm happy to talk it through more and hear your reasoning 😄

**Comment by killev** [comment]
Hi, 

Sorry for the delay.
Let's think about this problem from the abstract view:
1. You have abstract stream (or signals no matter how to call) wich transfer some abstract data.
2. You have an abstract "collectionView" which represents collection based data. 
3. You have abstract ChangeSet - collection-based data which contains:
- Abstract Collection - current state of the collection
- Abstract Diff  - number of operation we have to do to with the previous state of the collection in order to get current.

If 1 satisfies 3 it can be bound to 2. And it's no matter how we get this ChangeSet. From batchUpdate, or from .diff() operator or custom logic - bind should support it all as this is described in the contract.

At the moment I don't understand why we need both patch and Diff. 

So let'sjust  think about actions which diff can appy:
1. reload (this operation just replaces all items in collectionView)
2. update (this operation inserts/removes/updates items in collectionView in order to reflect the new state  of collection)
3. empty - no operations.

1 and 2 don't intersect, so it's rather enum: 

enum  Diff {
case reload
case update(deletes:[], inserts:[], moves[], ... )
}

"empty" - is sub-option of "update" when there're no operations which can be ignored in bind's implementation. But still "empty" means to do nothing. It's importand as you provide API which should be extendable and as clear as possible.

Please keep in mind that Diff is NOT A STATE. This is just a set of operations between states. We shouldn't keep it anywhere. We should keep the collection only.

So at the moment I just offer to continue the discussion, as have to verify some swift's features we can use to simplify these abstractions.

**Comment by killev** [comment]
Hi,

What if we just introduce `var shouldRload:Bool` property and `init(collection: Collection)` which   to ChangesetProtocol:
```swift
public protocol ChangesetProtocol {
    ...
    var shouldReload: Bool { get }
    // Create a changeset for the given collection which reloads data.
    init(collection: Collection)   
    ...
}
```
And then Implement `class Changeset` like below:
```swift
open class Changeset<Collection, Operation, Diff: Instantiatable>: ChangesetProtocol {
    ...
    public required init(collection: Collection) {
        self.collection = collection
    }

    public var shouldReload: Bool {
        return precalculatedDiff == nil && precalculatedPatch == nil
    }
    ...
}
```
Also remove `public init(collection: Collection, patch: [Operation], diff: Diff) {` which accepts both patch and diff as unnecessasy
